[Dynamo] Automatically in-graph traceable tensor subclass ctors

mlazos opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/114389

Previously, dynamo would attempt to trace through the `__init__` of traceable tensor subclasses, since their constructors are AOT dispatcher traceable by definition, dynamo should automatically put these in the graph like we do for any other tensors. Not doing this is difficult because dynamo would need to apply mutations post tensor subclass creation in the graph. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135151



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

mlazos committed:
[Dynamo] Automatically in-graph traceable tensor subclass ctors

[ghstack-poisoned]

bdhirsh commented on torch/utils/_python_dispatch.py:
```diff
@@ -391,6 +391,11 @@ def is_traceable_wrapper_subclass(t: object) -> TypeGuard[TensorWithFlatten]:
         and hasattr(t, "__tensor_unflatten__")
     )
 
+def is_traceable_wrapper_subclass_type(t: Type) -> TypeGuard[Type[TensorWithFlatten]]:
```
output of this fn should be `-> bool:` right?

bdhirsh replied:
(oh same with the one above but I guess that is old)

bdhirsh replied:
also nit: can you just update `is_traceable_wrapper_subclass` to `return is_traceable_wrapper_subclass_type(type(x))` to dedup the code?

mlazos replied:
Oh I thought TypeGuard is just a special bool, right?

mlazos commented on test/dynamo/test_subclasses.py:
```diff
@@ -1427,6 +1428,62 @@ def __metadata_guard__(self, x, y):
         )
 
+    def test_subclass_constructor_proxying(self):
+        class SubclassTensor(torch.Tensor):
```
Yup, once the other PR merges I'll test both these cases. Good idea!

bdhirsh reviewed (approved):
lgtm minus nits!

mlazos committed:
Update on "[Dynamo] Automatically in-graph traceable tensor subclass ctors"

Fixes https://github.com/pytorch/pytorch/issues/114389

Previously, dynamo would attempt to trace through the `__init__` of traceable tensor subclasses, since their constructors are AOT dispatcher traceable by definition, dynamo should automatically put these in the graph like we do for any other tensors. Not doing this is difficult because dynamo would need to apply mutations post tensor subclass creation in the graph. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge

mlazos committed:
Update on "[Dynamo] Automatically in-graph traceable tensor subclass ctors"

Fixes https://github.com/pytorch/pytorch/issues/114389

Previously, dynamo would attempt to trace through the `__init__` of traceable tensor subclasses, since their constructors are AOT dispatcher traceable by definition, dynamo should automatically put these in the graph like we do for any other tensors. Not doing this is difficult because dynamo would need to apply mutations post tensor subclass creation in the graph. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge

mlazos committed:
Update on "[Dynamo] Automatically in-graph traceable tensor subclass ctors"

Fixes https://github.com/pytorch/pytorch/issues/114389

Previously, dynamo would attempt to trace through the `__init__` of traceable tensor subclasses, since their constructors are AOT dispatcher traceable by definition, dynamo should automatically put these in the graph like we do for any other tensors. Not doing this is difficult because dynamo would need to apply mutations post tensor subclass creation in the graph. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

mlazos commented:
@pytorchbot merge
